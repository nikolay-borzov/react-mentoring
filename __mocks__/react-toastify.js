import * as toastify from 'react-toastify'

toastify.toast.error = jest.fn()

export * from 'react-toastify'
