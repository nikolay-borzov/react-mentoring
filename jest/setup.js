import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
