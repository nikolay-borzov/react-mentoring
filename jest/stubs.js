export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
