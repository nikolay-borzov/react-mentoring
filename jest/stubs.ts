import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
