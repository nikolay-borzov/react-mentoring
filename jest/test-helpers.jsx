import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
