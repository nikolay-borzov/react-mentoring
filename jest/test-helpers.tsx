import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
