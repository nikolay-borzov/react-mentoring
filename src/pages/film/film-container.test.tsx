import React from 'react'
import { shallow } from 'enzyme'
import { toast } from 'react-toastify'

import { film, films } from '../../../jest/stubs'
import { itRendersCorrectlyShallow } from '../../../jest/test-helpers'

import FilmContainerConnected, {
  FilmContainer,
  FilmContainerProps
} from './film-container'
import configureStore from '../../redux/create'

jest.mock('../../services/film-service')

describe('FilmContainer page component', () => {
  const fetchFilmMock = jest.fn()
  const reFetchFilmsMock = jest.fn()

  let props: FilmContainerProps = {
    match: {
      params: {
        id: film.id.toString()
      }
    },
    film: undefined,
    filmIsFetching: false,
    filmError: undefined,
    genre: 'Horror',
    relatedFilms: films,
    relatedFilmsIsFetching: false,
    relatedFilmsError: undefined,
    fetchFilm: fetchFilmMock,
    reFetchFilms: reFetchFilmsMock
  }

  const render = () => {
    const wrapper = shallow<FilmContainer>(<FilmContainer {...props} />)

    return { instance: wrapper.instance(), wrapper }
  }

  beforeEach(() => {
    fetchFilmMock.mockClear()
    reFetchFilmsMock.mockClear()

    props.film = film
    props.filmError = undefined
    props.relatedFilms = films
    props.relatedFilmsError = undefined
  })

  afterEach(() => {
    global.IS_SERVER = false
  })

  describe('renders correctly', () => {
    // TODO: Disabled due to type error or passing store as prop isn't working anymore
    /* itRendersCorrectlyShallow(() => {
      const { store } = configureStore()
      return <FilmContainerConnected store={store} />
    }, 'when connected to the store') */

    itRendersCorrectlyShallow(
      () => <FilmContainer {...props} />,
      'when film and relatedFilms are provided'
    )

    itRendersCorrectlyShallow(() => {
      props.film = undefined
      return <FilmContainer {...props} />
    }, 'when the film is unavailable')

    itRendersCorrectlyShallow(() => {
      props.film = {} as any
      return <FilmContainer {...props} />
    }, 'when the film is not found')

    itRendersCorrectlyShallow(() => {
      props.relatedFilms = []
      return <FilmContainer {...props} />
    }, 'when no related films found')

    itRendersCorrectlyShallow(() => {
      props.relatedFilms = []
      props.relatedFilmsError = new Error('Load error')
      return <FilmContainer {...props} />
    }, 'when related films had failed to load')

    itRendersCorrectlyShallow(() => {
      props.film = undefined
      props.relatedFilms = []
      return <FilmContainer {...props} />
    }, 'when no film found nor related films found ')
  })

  it('loads film on sever side', () => {
    global.IS_SERVER = true

    render()

    expect(fetchFilmMock).toHaveBeenCalledWith(film.id)
  })

  it('re-fetches films if film already loaded', () => {
    props.film = film

    render()

    expect(reFetchFilmsMock).toHaveBeenCalled()
  })

  it(`loads film on client side if it isn't loaded yet`, () => {
    props.film = undefined

    render()

    expect(fetchFilmMock).toHaveBeenCalledWith(film.id)
  })

  it('loads a film if id has changed', () => {
    const { wrapper, instance } = render()
    // Clear initial call
    fetchFilmMock.mockClear()

    wrapper.setProps({ match: { params: { id: (film.id + 1).toString() } } })
    // https://github.com/airbnb/enzyme/issues/34
    instance.componentDidUpdate(props)

    expect(fetchFilmMock).toHaveBeenCalled()
  })

  it(`doesn't load a film if id hasn't changed`, () => {
    const { wrapper, instance } = render()
    fetchFilmMock.mockClear()

    wrapper.setProps({ genre: 'Drama' })
    instance.componentDidUpdate(props)

    expect(fetchFilmMock).not.toHaveBeenCalled()
  })

  it('displays an error when unable to load the film', () => {
    const error = new Error('Film load error')

    const { wrapper, instance } = render()

    wrapper.setProps({ filmError: error })
    instance.componentDidUpdate(props)

    expect(toast.error).toHaveBeenCalled()
  })

  it('displays error when unable to load related films', () => {
    const error = new Error('Films load error')

    const { wrapper, instance } = render()

    wrapper.setProps({ relatedFilmsError: error })
    instance.componentDidUpdate(props)

    expect(toast.error).toHaveBeenCalled()
  })
})
