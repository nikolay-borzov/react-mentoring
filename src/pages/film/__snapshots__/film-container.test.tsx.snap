// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmContainer page component renders correctly when film and relatedFilms are provided 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Film title
       ðŸŽ¥ Movie Search
    </title>
  </HelmetWrapper>
  <Header>
    <div
      className="padding-controls flex flex-align-center"
    >
      <div
        className="flex-grow"
      >
        <SiteName />
      </div>
      <NavigateButton
        primary={true}
        small={true}
        to="/search"
      >
        Search
      </NavigateButton>
    </div>
    <LoadingBlock
      hideText={true}
      isLoaded={true}
    >
      <FilmDetails
        film={
          Object {
            "budget": 15000000,
            "genres": Array [
              "some genre",
              "another genre",
            ],
            "id": 1,
            "overview": "Once upon a time...",
            "poster_path": "https://picsum.photos/g/300/450/?random",
            "release_date": "2018-05-04",
            "revenue": 59735548,
            "tagline": "Put that cookie down",
            "title": "Film title",
            "vote_average": 7.5,
            "vote_count": 667,
          }
        }
      />
    </LoadingBlock>
  </Header>
  <main
    className="content"
  >
    <LoadingBlock
      hideText={false}
      isLoaded={true}
    >
      <SearchResultsPanel>
        Films by 
        Horror
         genre
      </SearchResultsPanel>
      <LoadableComponent
        films={
          Array [
            Object {
              "budget": 15000000,
              "genres": Array [
                "some genre",
                "another genre",
              ],
              "id": 1,
              "overview": "Once upon a time...",
              "poster_path": "https://picsum.photos/g/300/450/?random",
              "release_date": "2018-05-04",
              "revenue": 59735548,
              "tagline": "Put that cookie down",
              "title": "Film title",
              "vote_average": 7.5,
              "vote_count": 667,
            },
          ]
        }
      />
    </LoadingBlock>
  </main>
  <Footer />
</Fragment>
`;

exports[`FilmContainer page component renders correctly when no film found nor related films found  1`] = `
<Fragment>
  <Header>
    <div
      className="padding-controls flex flex-align-center"
    >
      <div
        className="flex-grow"
      >
        <SiteName />
      </div>
      <NavigateButton
        primary={true}
        small={true}
        to="/search"
      >
        Search
      </NavigateButton>
    </div>
    <LoadingBlock
      hideText={true}
      isLoaded={true}
    >
      <FilmDetails />
    </LoadingBlock>
  </Header>
  <main
    className="content"
  >
    <LoadingBlock
      hideText={false}
      isLoaded={true}
    >
      <SearchResultsPanel>
        Films by 
        Horror
         genre
      </SearchResultsPanel>
    </LoadingBlock>
  </main>
  <Footer />
</Fragment>
`;

exports[`FilmContainer page component renders correctly when no related films found 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Film title
       ðŸŽ¥ Movie Search
    </title>
  </HelmetWrapper>
  <Header>
    <div
      className="padding-controls flex flex-align-center"
    >
      <div
        className="flex-grow"
      >
        <SiteName />
      </div>
      <NavigateButton
        primary={true}
        small={true}
        to="/search"
      >
        Search
      </NavigateButton>
    </div>
    <LoadingBlock
      hideText={true}
      isLoaded={true}
    >
      <FilmDetails
        film={
          Object {
            "budget": 15000000,
            "genres": Array [
              "some genre",
              "another genre",
            ],
            "id": 1,
            "overview": "Once upon a time...",
            "poster_path": "https://picsum.photos/g/300/450/?random",
            "release_date": "2018-05-04",
            "revenue": 59735548,
            "tagline": "Put that cookie down",
            "title": "Film title",
            "vote_average": 7.5,
            "vote_count": 667,
          }
        }
      />
    </LoadingBlock>
  </Header>
  <main
    className="content"
  >
    <LoadingBlock
      hideText={false}
      isLoaded={true}
    >
      <SearchResultsPanel>
        Films by 
        Horror
         genre
      </SearchResultsPanel>
      <LoadableComponent
        className="font-size-header font-bold color-alt"
      >
        No related movies found
      </LoadableComponent>
    </LoadingBlock>
  </main>
  <Footer />
</Fragment>
`;

exports[`FilmContainer page component renders correctly when related films had failed to load 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Film title
       ðŸŽ¥ Movie Search
    </title>
  </HelmetWrapper>
  <Header>
    <div
      className="padding-controls flex flex-align-center"
    >
      <div
        className="flex-grow"
      >
        <SiteName />
      </div>
      <NavigateButton
        primary={true}
        small={true}
        to="/search"
      >
        Search
      </NavigateButton>
    </div>
    <LoadingBlock
      hideText={true}
      isLoaded={true}
    >
      <FilmDetails
        film={
          Object {
            "budget": 15000000,
            "genres": Array [
              "some genre",
              "another genre",
            ],
            "id": 1,
            "overview": "Once upon a time...",
            "poster_path": "https://picsum.photos/g/300/450/?random",
            "release_date": "2018-05-04",
            "revenue": 59735548,
            "tagline": "Put that cookie down",
            "title": "Film title",
            "vote_average": 7.5,
            "vote_count": 667,
          }
        }
      />
    </LoadingBlock>
  </Header>
  <main
    className="content"
  >
    <LoadingBlock
      hideText={false}
      isLoaded={true}
    >
      <SearchResultsPanel>
        Films by 
        Horror
         genre
      </SearchResultsPanel>
      <LoadableComponent
        className="error-message"
      >
        Unable to load related movies
      </LoadableComponent>
    </LoadingBlock>
  </main>
  <Footer />
</Fragment>
`;

exports[`FilmContainer page component renders correctly when the film is not found 1`] = `<LoadableComponent />`;

exports[`FilmContainer page component renders correctly when the film is unavailable 1`] = `
<Fragment>
  <Header>
    <div
      className="padding-controls flex flex-align-center"
    >
      <div
        className="flex-grow"
      >
        <SiteName />
      </div>
      <NavigateButton
        primary={true}
        small={true}
        to="/search"
      >
        Search
      </NavigateButton>
    </div>
    <LoadingBlock
      hideText={true}
      isLoaded={true}
    >
      <FilmDetails />
    </LoadingBlock>
  </Header>
  <main
    className="content"
  >
    <LoadingBlock
      hideText={false}
      isLoaded={true}
    >
      <SearchResultsPanel>
        Films by 
        Horror
         genre
      </SearchResultsPanel>
      <LoadableComponent
        films={
          Array [
            Object {
              "budget": 15000000,
              "genres": Array [
                "some genre",
                "another genre",
              ],
              "id": 1,
              "overview": "Once upon a time...",
              "poster_path": "https://picsum.photos/g/300/450/?random",
              "release_date": "2018-05-04",
              "revenue": 59735548,
              "tagline": "Put that cookie down",
              "title": "Film title",
              "vote_average": 7.5,
              "vote_count": 667,
            },
          ]
        }
      />
    </LoadingBlock>
  </main>
  <Footer />
</Fragment>
`;
